### DISCUSSION TITLE
Require confirmation for photo component

### DISCUSSION THREAD

**Discussion by mixmaniax** [Budibase/budibase#7587] [Help] [answered]
Hi there
Is there a way to require a confirmation before deleting a photo/picture using the photo component on a form. I would like to avoid that someone by mistake deletes it when pressing the X - as it will immediately delete the shown picture.

Thank you in advance for any suggestion.

**Comment by melohagan** [comment]
Hi @mixmaniax 

I'm not quite sure which component you are referring to when you say "photo component" ?

**Reply by mixmaniax** [reply]
Hi @melohagan 

Sorry, I mean the attachment component...see screenshot below...

<img width="681" alt="Bildschirmfoto 2022-09-12 um 16 06 32" src="https://user-images.githubusercontent.com/2330603/189675297-be46ced7-663a-440a-a077-0cebe9ea6faf.png">

**Reply by melohagan** [reply]
Perfect, thank you. There's no way to do this at the moment but it might be something we can add. I'm just wondering what the issue is. If they click the 'X' can they not just upload the image again?

**Reply by mixmaniax** [reply]
Thank you @melohagan 

Indeed re-upload would be an option. The issue in my use case is, that I'm thinking of a new employee onboarding process where the HR dept needs multiple information and also an ID or Passport picture along the info his/her manager need to provide for the application. I tested the attachment component with a smartphone and there is the option to take a picture directly from the smartphone. At that point once uploaded with the form obviously there is no local attachment for it. So in case someone of the HR dept. by mistake clicks on the "delete attachment" icon, that document/picture is gone without warning. That will generate additional workload for the HR/Manager and Employee to collect a copy of that document again. Having at least the possibility to activate a "require confirmation" for the delete action of the attachment component would surely reduce unnecessary administrative work.

Hope I was able to explain myself...

**Comment by melohagan** [comment, marked as answer]
Thanks for the additional context @mixmaniax 

I've come up with a solution that may work for you, you'd just need to style it to your liking:
![upload_warning](https://user-images.githubusercontent.com/101575380/189697911-7aaaba24-1068-453a-9733-0611ba7fd818.gif)

### Step 1 - Remove the clear attachment button i.e. the 'X'
Add an embed to the top of your screen with the following content:
```html
<style>
.delete-button {
  display: none !important;
}
</style>
```

### Step 2 - Clear Attachments button
Add a button to your form and give it a name. 
**Define actions** for on click and *Update State*; setting a variable named *clear* to 'true'.
<img width="920" alt="Screenshot 2022-09-12 at 16 48 55" src="https://user-images.githubusercontent.com/101575380/189698791-dd9e661d-d52c-4eea-9c84-21e75130a2db.png">

Also configure conditions for the button as follows:
<img width="920" alt="Screenshot 2022-09-12 at 16 49 27" src="https://user-images.githubusercontent.com/101575380/189698899-7fd7f37c-25ad-4308-a808-2f97311a5543.png">

*JavaScript binding*
```js
return (!$("State.clear") || $("State.clear") === "false") && $("New Form.Fields.Attachment")?.length > 0;
```
We are checking to see if if the Clear Attachments button has been clicked, but also we don't want to show the button if there are no attachments uploaded.

### Step 3 - Are you sure prompt
Add a container with a warning message and two buttons for Yes/No.
Your component tree should look roughly like so:
<img width="1227" alt="Screenshot 2022-09-12 at 16 51 56" src="https://user-images.githubusercontent.com/101575380/189699470-6f0cbbd9-10b1-40b9-8cfc-a4e5dc41d54f.png">

Add a condition to the **container** as follows:
<img width="915" alt="Screenshot 2022-09-12 at 16 53 11" src="https://user-images.githubusercontent.com/101575380/189699750-8e74f753-eb7e-411f-bcbc-a3a7fa573c80.png">

For the **Yes** button, define two actions as follows:
<img width="915" alt="Screenshot 2022-09-12 at 16 54 50" src="https://user-images.githubusercontent.com/101575380/189700110-5f9f69d6-1860-4c4e-9648-8ec38f394d70.png">
<img width="915" alt="Screenshot 2022-09-12 at 16 55 00" src="https://user-images.githubusercontent.com/101575380/189700148-4fcf2082-f80e-4d75-9d99-d60543d3a217.png">

Finally for the **Cancel** button, define the same action as above for setting the *clear* state to 'false'.

**Reply by mixmaniax** [reply]
Hi @melohagan 

Wow - thank you for the reply - looks promising. I followed your instruction, but I'm not able to make it work. The button in Step 2 - Clear Attachments button does not even shows up - seems to be something with the condition java. I used the "Employee Directory" template as to enhance with your instructions. Maybe I should start from scratch and try. Eventuelly could you export your app so I can import it and try it out?

Thank you very much for your very appreciated assistance.

**Reply by melohagan** [reply]
Maybe for step 2 your syntax is a little off or something isn't configured just right? It was a little fiddly because you have to check if the *clear* state is "false" or not set at all, because initially all app state variables are **undefined**.

Here is the app export; hope it helps! 
[Sales App-export-1663080374950.txt](https://github.com/Budibase/budibase/files/9558228/Sales.App-export-1663080374950.txt)

**Reply by mixmaniax** [reply]
Thank you very much for the example which works! 
Found why my example did not work (typo on my side with the form name in the javascript binding in step 2, sorry for that). 
What I now realize is that your solution would not be suitable in my use case because of the following:
1. for a row/record it will delete all attachments uploaded for an employee - e.g. employee "John"  uploaded three documents and we want to delete number 2 of 3 and not all of them 
2. clicking the "yes, clear the attachments!" will also clear out any form field e.g. the name of the employee and would need to be retyped along with all other form fields...

So I honestly guess for my use case there won't be a easy workarounds - if you have any other idea I'm happy to hear ;-)

Thank you anyhow for your effort in assisting my so quick!

**Reply by melohagan** [reply]
You're welcome. Instead of clearing the form, you can use the **Update Field Value** action to clear only the attachment field: https://docs.budibase.com/docs/actions#update-field-value

In addition, you could use JavaScript to *shift* the array if you only want to remove one attachment at a time. 
https://reactgo.com/js-remove-first-element-of-array/#removing-the-first-element
